core-plugins/src/main/java/io/cdap/plugin/batch/connector/FileConnector.java

```diff
@@ -0,0 +1,144 @@
+  private static final String FILE = "file";
+
+  @Override
+  public BrowseDetail browse(ConnectorContext connectorContext, BrowseRequest request) throws IOException {
+    String path = request.getPath();
+    File file = new File(request.getPath());
+    if (!file.exists()) {
+      throw new IllegalArgumentException(String.format("The given path %s does not exist", path));
+    }
+
+    // if it is not a directory, then it is not browsable, return the path itself
+    if (!file.isDirectory()) {
+      return BrowseDetail.builder().setTotalCount(1).addEntity(generateBrowseEntity(file)).build();
+    }
+
+    // list the files and classify them with file and directory
+    File[] files = file.listFiles();
+    if (files == null) {
+      throw new IOException(String.format("Unable to browse the path %s", path));
+    }
+
+    // sort the files by names
+    Arrays.sort(files);
+    int limit = request.getLimit() == null ? files.length : Math.min(request.getLimit(), files.length);
+    BrowseDetail.Builder builder = BrowseDetail.builder().setTotalCount(files.length);
+    for (int i = 0; i < limit; i++) {
+      // do not browse hidden files
+      if (files[i].isHidden()) {
+        continue;
+      }
+      builder.addEntity(generateBrowseEntity(files[i]));
+    }
+    return builder.build();
+  }
+
+  @Override
+  protected void setConnectorSpec(ConnectorSpecRequest request, ConnectorSpec.Builder builder) {
+    super.setConnectorSpec(request, builder);
+    builder.addRelatedPlugin(new PluginSpec(FileBatchSource.NAME, BatchSource.PLUGIN_TYPE,
+                                            Collections.singletonMap("path", request.getPath())));
+  }
+
+  private BrowseEntity generateBrowseEntity(File file) throws IOException {
+    String path = file.getCanonicalPath();
+    boolean isDirectory = file.isDirectory();
+    BrowseEntity.Builder builder = BrowseEntity.builder(file.getName(), path, isDirectory ? DIRECTORY : FILE);
+    Set<BrowseEntityProperty> properties = new HashSet<>();
+
+    if (isDirectory) {
+      builder.canBrowse(true).canSample(true);
+    } else {
+      String fileType = FileTypeDetector.detectFileType(path);
+      builder.canSample(FileTypeDetector.isSampleable(fileType));
+      properties.add(BrowseEntityProperty.builder(FILE_TYPE_KEY, fileType,
+                                                  BrowseEntityProperty.PropertyType.STRING).build());
+      properties.add(BrowseEntityProperty.builder(SIZE_KEY, String.valueOf(file.length()),
+                                                  BrowseEntityProperty.PropertyType.SIZE_BYTES).build());
+    }
+
+    properties.add(BrowseEntityProperty.builder(LAST_MODIFIED_KEY, String.valueOf(file.lastModified()),
+                                                BrowseEntityProperty.PropertyType.TIMESTAMP_MILLIS).build());
+    properties.add(BrowseEntityProperty.builder(OWNER_KEY, Files.getOwner(file.toPath()).getName(),
```

Comment:
is it possible that Files.getOwner return null ?

Reply:
doesn't hurt to protect against possible NPE if we're unsure

Reply:
The method is not marked as Nullable, the javadoc also doesn't mention it, so I assume it won't return null?